prime-angular-frontend/src/app/modules/site-registration/pages/device-provider-page/device-provider-page.component.ts

```diff
@@ -0,0 +1,176 @@
+    this.currentIndex = -1;
+  }
+
+  public onBack() {
+    const nextRoute = (this.isCompleted)
+      ? SiteRoutes.SITE_REVIEW
+      : SiteRoutes.HOURS_OPERATION;
+
+    this.routeUtils.routeRelativeTo(nextRoute);
+  }
+
+  public onCancel() {
+    this.isEditing = false;
+  }
+
+  public onAdd(): void {
+    this.currentIndex = -1;
+    this.form = this.formState.buildIndividualDeviceProvider();
+    this.isEditing = true;
+  }
+
+  public onEdit(index: number): void {
+    this.currentIndex = index;
+    this.form = this.formState.at(index);
+    this.isEditing = true;
+  }
+
+  public onContinue(): void {
+    if (this.formUtilsService.checkValidity(this.form)) {
+      if (this.currentIndex !== -1) {
+        this.formState.individualDeviceProviders.at(this.currentIndex).patchValue(this.form);
+      } else {
+        this.formState.individualDeviceProviders.push(this.form);
+      }
+      this.updateCardList(this.formState.json);
+      this.isEditing = false;
+    }
+  }
+
+  public onRemove(index: number): void {
+    this.formState.individualDeviceProviders.removeAt(index);
+    this.updateCardList(this.formState.json);
+  }
+
+  ngOnInit(): void {
+    this.createFormInstance();
+    this.patchForm();
+  }
+
+  protected createFormInstance() {
+    this.formState = this.siteFormStateService.deviceProviderFormState;
+    this.form = this.formState.buildIndividualDeviceProvider();
+  }
+
+  protected patchForm(): void {
+    this.site = this.siteService.site;
+    this.isCompleted = this.site?.completed;
+    this.siteFormStateService.setForm(this.site, !this.hasBeenSubmitted);
+    this.updateCardList(this.formState.json);
+    this.isEditing = !this.formState.json.length;
+  }
+
+  /**
+   * @description
+   * Update the contact card list.
+   */
+  private updateCardList(deviceProviders: IndividualDeviceProvider[]): void {
+    this.contactCardListItems = this.getCardList(deviceProviders);
+
+    if (!deviceProviders.length) {
+      this.isEditing = true;
+      this.form = this.formState.buildIndividualDeviceProvider();
+    }
+  }
+
+  private getCardList(deviceProviders: IndividualDeviceProvider[]): CardListItem[] {
+    return deviceProviders.map(this.getListItem());
+  }
+
+  protected getListItem(): (deviceProvider: IndividualDeviceProvider) => CardListItem {
+    return (deviceProvider: IndividualDeviceProvider) => ({
+      icon: 'account_circle',
+      title: `${deviceProvider.firstName} ${deviceProvider.lastName}`,
+      properties: [
+        { key: 'Date of Birth', value: this.formatDatePipe.transform(deviceProvider.dateOfBirth) },
+        { key: 'Email', value: deviceProvider.email }
+      ],
+      action: {
+        title: 'Update Information'
+      }
+    });
+  }
+
+  protected additionalValidityChecks(obj: { individualDeviceProviders: IndividualDeviceProvider[] }): boolean {
+    // At least one provider needs to exist
+    return (obj.individualDeviceProviders.length > 0);
```

Comment:
You won't need the parenthesis since length will either be 0 (falsey) or greater than 0 (truthy)

Reply:
You'll find `Pick<...>` effective for communicating the params and knowing where they originate and in your case you don't have a `IndividualDeviceProviderForm` model so you could use `Site`

```
protected additionalValidityChecks({ individualDeviceProviders }: Pick<Site, 'individualDeviceProviders'>): boolean {
  return individualDeviceProviders?.length;
}
```